src/proxy/src/main/java/ee/ria/xroad/proxy/clientproxy/ClientRestMessageHandler.java

```diff
@@ -92,17 +119,59 @@ public void sendErrorResponse(HttpServletResponse response, CodedException ex) t
         } else {
             response.setStatus(HttpStatus.BAD_REQUEST_400);
         }
-        response.setContentType("application/json");
         response.setCharacterEncoding(MimeUtils.UTF8);
         response.setHeader("X-Road-Error", ex.getFaultCode());
 
-        final JsonWriter writer = new JsonWriter(new PrintWriter(response.getOutputStream()));
-        writer.beginObject()
-                .name("type").value(ex.getFaultCode())
-                .name("message").value(ex.getFaultString())
-                .name("detail").value(ex.getFaultDetail())
-                .endObject()
-                .close();
+        String responseContentType = decideErrorResponseContentType(requestAcceptType);
+        response.setContentType(responseContentType);
+        if (XML_TYPES.contains(responseContentType)) {
+            DocumentBuilderFactory docFactory = XmlUtils.createDocumentBuilderFactory();
+            try {
+                DocumentBuilder docBuilder = docFactory.newDocumentBuilder();
+                Document doc = docBuilder.newDocument();
+                Element errorRootElement = doc.createElement("error");
+                doc.appendChild(errorRootElement);
+                Element typeElement = doc.createElement("type");
+                typeElement.appendChild(doc.createTextNode(ex.getFaultCode()));
+                errorRootElement.appendChild(typeElement);
+                Element messageElement = doc.createElement("message");
+                messageElement.appendChild(doc.createTextNode(ex.getFaultString()));
+                errorRootElement.appendChild(messageElement);
+                Element detailElement = doc.createElement("detail");
+                detailElement.appendChild(doc.createTextNode(ex.getFaultDetail()));
+                errorRootElement.appendChild(detailElement);
+                response.getOutputStream().write(XmlUtils.prettyPrintXml(doc, "UTF-8", 0).getBytes());
+            } catch (Exception e) {
+                log.error("Unable to generate XML document");
+            }
+        } else {
+            final JsonWriter writer = new JsonWriter(new PrintWriter(response.getOutputStream()));
+            writer.beginObject()
+                    .name("type").value(ex.getFaultCode())
+                    .name("message").value(ex.getFaultString())
+                    .name("detail").value(ex.getFaultDetail())
+                    .endObject()
+                    .close();
+        }
     }
 
+    private String decideErrorResponseContentType(String acceptHeaderValue) {
+        List<String> allPieces = new ArrayList<>();
+        String[] commaSplit = acceptHeaderValue.trim().split("\\s*,\\s*");
+        for (String s1: commaSplit) {
+            String[] colonSplit = s1.trim().split("\\s*;\\s*");
+            for (String s2: colonSplit) {
+                allPieces.add(s2);
+            }
+        }
+        if (allPieces.stream().anyMatch(TEXT_XML::equalsIgnoreCase)) {
```

Comment:
if..else..else.. could be shortened to (assuming media type is normalized to lower-case previously)
`return allPieces.stream().filter(XML_TYPES::contains).findAny().orElse(APPLICATION_JSON)`

Reply:
It doesn't handle `text/*` correctly (should return `text/xml` content type)

Reply:
Add a tranformation for that:
allPieces.stream().filter(XML_TYPES::contains)
.findAny()
.map("text/* to text/xml, else original")
.orElse(APPLICATION_JSON)

Reply:
Ok, updated